hail/src/main/scala/is/hail/expr/types/physical/PDict.scala

```diff
@@ -1,24 +1,26 @@
 package is.hail.expr.types.physical
 
-import is.hail.annotations.{UnsafeUtils, _}
-import is.hail.check.Gen
+import is.hail.annotations._
 import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.virtual.TDict
-import is.hail.utils._
-import org.json4s.jackson.JsonMethods
 
-import scala.reflect.{ClassTag, _}
-
-final case class PDict(keyType: PType, valueType: PType, override val required: Boolean = false) extends PContainer {
-  lazy val virtualType: TDict = TDict(keyType.virtualType, valueType.virtualType, required)
+object PDict {
+  def apply(keyType: PType, valueType: PType, required: Boolean): PDict = PDict(new PDictStruct(keyType, valueType), required)
+  def apply(keyType: PType, valueType: PType): PDict = PDict(keyType, valueType, false)
+}
 
-  val elementType: PStruct = PStruct(required = true, "key" -> keyType, "value" -> valueType)
+class PDictStruct(val keyType: PType, val valueType: PType) {
```

Comment:
This is the cleanest way I found to do this, because I cannot pass an elementType that is the constructor (body of the class), to the type signature (`extend PArrayBackedContainer(elementType, required)` won't work, if `val elementType = ...` is defined in the body constructor). Is there something in Scala that allows this?

I played around doing this without PDictStruct, but this made the apply method defined by the case class implicitly a bit confusing. I would not get type check errors if someone did PDict(PStruct("blah" -> "foo")). This seemed relatively clean, and I also like that it reinforce that PDict is really doubly nested (PArray and PStruct)

Reply:
Here's a way to make that work:
```
final case class PDict(keyType: PType, valueType: PType, required: Boolean = false) extends PArrayBackedContainer(PCanonicalArray(PStruct(required = true, ("key", keyType), ("value", valueType), required)) {

val elementType: PStruct = arrayRep.elementType.asInstanceOf[PStruct]
```

I don't like the added cognitive overhead of PDictStruct.

Reply:
Sure